google/auth/credentials.py

```diff
@@ -0,0 +1,209 @@
+    construction. Some classes will provide mechanisms to copy the credentials
+    with modifications such as :meth:`ScopedCredentials.with_scopes`.
+    """
+    def __init__(self):
+        self.token = None
+        """str: The bearer token that can be used in HTTP headers to make
+        authenticated requests."""
+        self.expiry = None
+        """Optional[datetime]: When the token expires and is no longer valid.
+        If this is None, the token is assumed to never expire."""
+
+    @property
+    def valid(self):
+        """Checks the validity of the credentials.
+
+        This is True if the credentials have a :attr:`token` and the token
+        is not :attr:`expired`.
+        """
+        return self.token is not None and not self.expired
+
+    @property
+    def expired(self):
+        """Checks if the credentials are expired.
+
+        Note that credentials can be invalid but not expired becaue Credentials
+        with :attr:`expiry` set to None is considered to never expire.
+        """
+        now = _helpers.utcnow()
+        if self.expiry is None or self.expiry > now:
+            return False
+        else:
+            return True
+
+    @abc.abstractmethod
+    def refresh(self, request):
+        """Refreshes the access token.
+
+        Args:
+            request (google.auth.transport.Request): A callable used to make
+                HTTP requests.
+
+        Raises:
+            google.auth.exceptions.RefreshError: If the credentials could
+                not be refreshed.
+        """
+        # pylint: disable=missing-raises-doc
+        # (pylint doesn't recognize that this is abstract)
+        raise NotImplementedError('Refresh must be implemented')
+
+    def apply(self, headers, token=None):
+        """Apply the token to the authentication header.
+
+        Args:
+            headers (Mapping): The HTTP request headers.
+            token (Optional[str]): If specified, overrides the current access
+                token.
+        """
+        headers[b'authorization'] = 'Bearer {}'.format(
+            _helpers.from_bytes(token or self.token))
+
+    def before_request(self, request, method, url, headers):
+        """Performs credential-specific before request logic.
+
+        Refreshes the credentials if necessary, then calls :meth:`apply` to
+        apply the token to the authentication header.
+
+        Args:
+            request (google.auth.transport.Request): A callable used to make
+                HTTP requests.
+            method (str): The request's HTTP method.
+            url (str): The request's URI.
+            headers (Mapping): The request's headers.
+        """
+        # pylint: disable=unused-argument
+        # (Subclasses may use these arguments to ascertain information about
+        # the http request.)
+        if not self.valid:
+            self.refresh(request)
+        self.apply(headers)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ScopedCredentials(object):
```

Comment:
Ahh. Aha! This might be the best place to attempt [one of the concrete recommendations for which you asked](https://github.com/google/oauth2client/issues/597#issuecomment-252756551).

What's confusing me and worrying me about this class right now is that it is a public API element that appears intended to serve both internal and external audiences - its documentation clearly speaks to developer-users of the library, but it also has private and double-private code elements intended for use by other parts of the library. The single-underscore code elements threw me because (1) I only ever use single-underscore to mean private-to-a-code-element (the class, in this case) and the single-underscore code elements they aren't enough on their own to add up to useful behavior within just the class, and (2) I'm aware that others use single-underscore to mean visible-to-subclasses-like-Java's-`protected` but the class isn't documented sufficiently for subclassing. Double-underscore I see rarely enough and it always seems to mean "really really private".

Reading the rendered documentation really helps this class feel more like a completely-abstract interface. **But:** this is open source, and the documentation includes links to the source (except for `requires_scopes`; that's weird), so I wonder: are developer-users of the library going to inherit from this class, call inherited concrete behavior, and get upset when that inherited concrete behavior changes? That's where I think there's accidental leakage - the class is in the library's public API, its methods are class's public API and therefore in the library API, and yet its methods have concrete implementations intended only for library-internal use and those concrete implementations are totally accessible through the API.

My specific recommendation would be to express this class as
```
@six.add_metaclass(abc.ABCMeta)
class ScopedCredentials(object):
    """Interface for scoped credentials.

    OAuth 2.0-based credentials allow limiting access using scopes as described
    in `RFC6749 Section 3.3`_.

    If a credential class implements this interface then the credentials either
    require or use scopes in their implementation.

    Some credentials have scopes but do not allow or require scopes to be set.

    You can check if scoping is necessary with :attr:`requires_scopes`::
        if credentials.requires_scopes:
            credentials = credentials.create_scoped(['one', 'two'])

    .. _RFC6749 Section 3.3: https://tools.ietf.org/html/rfc6749#section-3.3
    """

    @abc.abstractmethod
    def requires_scopes(self):
        """Return True if these credentials require scopes to obtain an access token.
        """
        raise NotImplementedError()

    @abc.abstractmethod
    def with_scopes(self, scopes):
        """Create a copy of these credentials with the specified scopes.

        Args:
            scopes (Union[str, Sequence]): The scope or list of scopes to
                request.

        Returns:
          A `ScopedCredentials` equal to this object but for the additional scopes.
        """
        raise NotImplementedError()

    @abc.abstractmethod
    def has_scopes(self, scopes):
        """Checks if the credentials have the given scopes.

        .. warning: This method is not guaranteed to be accurate if the
            credentials are :attr:`~Credentials.invalid`.

        Returns:
            bool: True if the credentials have the given scopes.
        """
        raise NotImplementedError()
```
.

I have a feeling that implementations of this interface only make sense if they are also implementations of `Credentials`. If that's the case, and `Credentials` also has its undocumented-but-accidentally-publicly-accessible concrete elements removed, then should `ScopedCredentials` subclass `Credentials`? It's fine for one interface to extend another, right?

Responding to "There's no part of the credentials that I expect to be implemented by the consuming application - implementations are all first party but the interfaces are for the benefit of the users": Over a long period of time I have had little success presenting a class in a public API, documenting that it is for me to subclass and for users to merely use, and having users respect that boundary and not subclass the class. That's part of what's led me to the place of no-classes-with-concrete-elements-in-my-APIs: if a class is a pure interface, users can subclass it to their heart's content without depending on any of my concrete behavior or getting bothered when my internals change. You don't support users subclassing `ScopedCredentials` in its current form, but your currently-drafted documentation doesn't expressly say "only classes in the library are allowed to subclass this class". Given that after my first reading of its code my reaction was "it looks like I'm supposed to subclass this but I don't see enough in the documentation about how to do so", how many others will react the same way?

Responding to "Everything in credentials is abstract and is intended to tell the developer what can be done with a specific credential instance": as currently drafted, users are free to subclass `ScopedCredentials`, to call methods like `requires_scopes`, and to build on that concrete behavior that is not supported for them.

A secondary concern, from a future maintainer's perspective: with all else held equal (particularly the number of code elements in the public API and their size), smaller total code elements and more of them are better than fewer, larger total code elements. You don't seem to have to privately partially implement your interface in the same code element (`class ScopedCredentials`) in which you publicly define your interface, so that alone is an argument against doing so.

If you're in the mood for levity, the idea that a code element (particularly a class) should be either _entirely in_ or _entirely out_ of the API of the component in which it is placed (rather than having a mix of some subelements included in the API and some subelements intended to be used only within the component) is what I've taken to calling "The Hokey Pokey API Design Principle". If you're not in the mood for levity, I hope it's still a good principle.